File: src/main/java/com/gandan/practice/semaphore/BlockingSemaphore.java
package com.gandan.practice.semaphore;

import java.util.concurrent.Semaphore;


public class BlockingSemaphore {
    Semaphore sem;

    public BlockingSemaphore(){
        sem = new Semaphore(0);
    }

    public void lock(){
         sem.acquireUninterruptibly();
    }

    public void unlock(){
         sem.release();
    }
}

File: src/main/java/com/gandan/practice/semaphore/BlockingSemaphore.kt
package com.gandan.practice.semaphore

import java.util.concurrent.Semaphore


class BlockingSemaphore {
    internal var sem: Semaphore

    init {
        sem = Semaphore(0)
    }

    fun lock() {
        sem.acquireUninterruptibly()
    }

    fun unlock() {
        sem.release()
    }
}
